registration/include/pcl/registration/impl/ndt.hpp

```diff
@@ -169,23 +167,46 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeDerivatives (Eige
                                                                             const Eigen::Matrix<double, 6, 1> &transform,
                                                                             bool compute_hessian)
 {
-  score_gradient.setZero ();
-  hessian.setZero ();
-  double score = 0;
+  std::vector<double> scores(num_threads_, 0.0);
+  std::vector<Eigen::Matrix<double, 6, 1>, Eigen::aligned_allocator<Eigen::Matrix<double, 6, 1>>> score_gradients(num_threads_, Eigen::Matrix<double, 6, 1>::Zero().eval());
+  std::vector<Eigen::Matrix<double, 6, 6>, Eigen::aligned_allocator<Eigen::Matrix<double, 6, 6>>> hessians(num_threads_, Eigen::Matrix<double, 6, 6>::Zero().eval());
 
   // Precompute Angular Derivatives (eq. 6.19 and 6.21)[Magnusson 2009]
   computeAngleDerivatives (transform);
 
   // Update gradient and hessian for each point, line 17 in Algorithm 2 [Magnusson 2009]
-  for (std::size_t idx = 0; idx < input_->size (); idx++)
+  #pragma omp parallel for num_threads(num_threads_) schedule(guided, 8)
+  for (int idx = 0; idx < static_cast<int>(input_->size ()); idx++)
   {
   // Transformed Point
     const auto& x_trans_pt = trans_cloud[idx];
 
+    // Initialize Point Jacobian and Hessian
+    Eigen::Matrix<double, 4, 6> point_jacobian = Eigen::Matrix<double, 4, 6>::Zero();
+    Eigen::Matrix<double, 24, 6> point_hessian = Eigen::Matrix<double, 24, 6>::Zero();
+    point_jacobian.block<3, 3>(0, 0).setIdentity();
+
     // Find neighbors (Radius search has been experimentally faster than direct neighbor checking.
     std::vector<TargetGridLeafConstPtr> neighborhood;
     std::vector<float> distances;
```

Comment:
Perhaps, move `distances` and `neighborhood` to outside the loop, with some heuristic size if possible
* clear them before using them
* declare them as privatefirst to OMP

Reply:
Done